spring-integration-file/src/main/java/org/springframework/integration/file/WatchServiceDirectoryScanner.java

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright 2015 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.integration.file;
+
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.FileSystems;
+import java.nio.file.FileVisitResult;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.nio.file.SimpleFileVisitor;
+import java.nio.file.StandardWatchEventKinds;
+import java.nio.file.WatchEvent;
+import java.nio.file.WatchKey;
+import java.nio.file.WatchService;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Collection;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+
+import org.springframework.context.SmartLifecycle;
+
+import reactor.core.support.Assert;
+
+/**
+ * Directory scanner that uses Java 7 {@link WatchService}.
+ *
+ * The initial state of the directory is collected during {@link #start()}. Subsequent
+ * polls return new files as reported by {@code ENTRY_CREATE} events.
+ * <p>
+ * While initially walking the directory, any subdirectories encountered are registered
+ * to watch for creation events.
+ * <p>
+ * If subdirectories are subsequentially added, they too are walked and registered for
+ * new creation events.
+ *
+ * @author Hezi Schrager
+ * @author Gary Russell
+ * @since 4.2
+ *
+ */
+public class WatchServiceDirectoryScanner extends DefaultDirectoryScanner implements SmartLifecycle {
```

Comment:
Even if we don't have build code quality checks, it would be better to mark this class with `@UsesJava7`.